src/GraphQL/Server.hs

```diff
@@ -287,94 +314,9 @@ instance forall typeName interfaces fields m.
-          -- TODO: See if we can prevent this from happening at compile time.
-          --
-          -- We should definitely not be panicking in our library code because
-          -- the user returned a Bool from their union type.
-          case GValue.toObject result of
-            Nothing -> panic $ "Expected object as result of union query: " <> show result
-            Just obj -> pure obj
-        else
-          runUnion @m @rest rh fragment
-  runUnion _ _ =
-    queryError "Non-InlineFragment used for a union type query."
-
--- TODO(jml): I don't understand why I can't extract (Object typeName interfaces fields)
--- from this instance as I did for the above instance.
-instance forall m typeName interfaces fields.
-         ( MonadIO m
-         , MonadThrow m
-         , RunFields m (RunFieldsType m fields)
-         , KnownSymbol typeName
-         , HasObjectDefinition (Object typeName interfaces fields)
-         ) => RunUnion m (Object typeName interfaces fields) where
-  runUnion lh fragment@(AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType queryTypeName) [] subSelection)) =
-    case getDefinition @(Object typeName interfaces fields) of
-      Left err -> queryError (AST.formatNameError err)
-      Right object ->
-        if getName object == queryTypeName
-        then do
-          result <- buildResolver @m @(Object typeName interfaces fields) lh subSelection
-          -- TODO: See if we can prevent this from happening at compile time.
-          case GValue.toObject result of
-            Nothing -> panic $ "Expected object as result of union query: " <> show result
-            Just obj -> pure obj
-        else
-          queryError ("Union type could not be resolved:" <> show fragment)
-  runUnion _ _ =
-    queryError "Non-InlineFragment used for a union type query."
-
-instance forall m ks ru.
-         ( MonadThrow m
-         , MonadIO m
-         , RunUnion m (RunUnionType m ru)
-         ) => HasGraph m (Union ks ru) where
-  type Handler m (Union ks ru) = RunUnionHandlerType m (RunUnionType m ru)
-  -- TODO: check sanity of query before executing it. E.g. we can't
-  -- have the same field name in two different fragment branches
-  -- (needs to take aliases into account).
-
-  -- query "{ ... on Human { name } }"
-  -- [SelectionInlineFragment (InlineFragment (NamedType "Human") [] [SelectionField (Field "" "name" [] [] [])])]
-  buildResolver handler selection = do
-    -- GraphQL invariant is that all items in a Union must be objects
-    -- which means 1) they have fields 2) They are ValueMap
-    values <- map GValue.unionObjects (traverse (runUnion @m @(RunUnionType m ru) handler) selection)
-    maybe (panic $ "Duplicate fields in values: " <> show values) (pure . GValue.ValueObject) values
+    r <- forM selectionSet (runFields @m @(RunFieldsType m fields) handler)
+    let (errs, fields) = foldr' (\(Result ea fa) (eb, fbs) -> (eb <> ea, fa:fbs)) ([], []) r
```

Comment:
As pointed out in chat, this is buggy and the sequence version is not. However, I'm always happy to eliminate inessential complexity in follow-up.